### ISSUE TITLE
Reproducible Mask

### ISSUE THREAD

**Issue by ywoelker** [ywoelker/LargeST#1]
We found that the mask should be reproducible over all baselines and executions. 
Therefore, we pregenerate masks for the input and output sparsity.
We generate the same amount of masks as we rerun the baselines to have different masks at each run. 

The mask type should be (samples x times x sensors) for point-wise and (samples x sensors) for spatial. 

We need to check with SPIN-H and OPCR publication. If they sample also epoch-wise.

**Comment by ywoelker** [comment]
Hey @Arashhs,

I removed with commit 6a3f7ad the experiment structure because it would not be possible to use it for persistent masks. 
Additionally, I added the new masks with commit de9c4c001. Let me briefly describe the new workflow. 

### Generating Masks

There is a script `data/generate_mask_for_experiments.py` which will create masks based on a bunch of dropout settings and save them in `data/masks/{dataset}/{folder_name}`. There will be a fixed set of pre-generated masks based on the `--n_masks ` option. This are the options in the script: 

```txt
usage: generate_mask_for_experiments.py [-h] --dataset DATASET [--input_dropout INPUT_DROPOUT] [--input_fault INPUT_FAULT] [--input_fault_minmax INPUT_FAULT_MINMAX INPUT_FAULT_MINMAX] [--input_sensor_dropout INPUT_SENSOR_DROPOUT] [--output_dropout OUTPUT_DROPOUT]
                                        [--output_fault OUTPUT_FAULT] [--output_fault_minmax OUTPUT_FAULT_MINMAX OUTPUT_FAULT_MINMAX] [--output_sensor_dropout OUTPUT_SENSOR_DROPOUT] [--share_input_output_mask SHARE_INPUT_OUTPUT_MASK] [--train_dropout TRAIN_DROPOUT]
                                        [--seed SEED] --experiment_folder EXPERIMENT_FOLDER [--n_masks N_MASKS]

options:
  -h, --help            show this help message and exit
  --dataset DATASET     Name of the dataset (e.g., "ca", "gba", "gla", "sd")
  --input_dropout INPUT_DROPOUT
                        Input dropout rate (between 0 and 1)
  --input_fault INPUT_FAULT
                        Input fault rate (between 0 and 1)
  --input_fault_minmax INPUT_FAULT_MINMAX INPUT_FAULT_MINMAX
                        Min and max length for input fault injection
  --input_sensor_dropout INPUT_SENSOR_DROPOUT
                        Input sensor dropout rate (between 0 and 1)
  --output_dropout OUTPUT_DROPOUT
                        Output dropout rate (between 0 and 1)
  --output_fault OUTPUT_FAULT
                        Output fault rate (between 0 and 1)
  --output_fault_minmax OUTPUT_FAULT_MINMAX OUTPUT_FAULT_MINMAX
                        Min and max length for output fault injection
  --output_sensor_dropout OUTPUT_SENSOR_DROPOUT
                        Output sensor dropout rate (between 0 and 1)
  --share_input_output_mask SHARE_INPUT_OUTPUT_MASK
                        Whether to have one mask for the complete dataset and take the X and y from the same dropouted data
  --train_dropout TRAIN_DROPOUT
                        Training sensor dropout rate (between 0 and 1)
  --seed SEED           Random seed for reproducibility
  --experiment_folder EXPERIMENT_FOLDER
                        Name of the experiment for the folder
  --n_masks N_MASKS     Number of different masks to generate
```

There will be an input and an output mask with the shape (total_samples x sensors), which is consistent with literature (OPCR, SPIN) and not as I first thought (total_samples x 12 x sensors). So overlapping samples experience the exact same missing values.

`input_dropout` - Just a normal dropout percentage
`input_sensor_dropout` - This removes complete sensors from the input and is inspired by the T4C22 challenge, which is referenced OPCR.
`input_fault` - This is done by the SPIN paper to imitate more realistic outtakes. First, it samples with a low percentage starting points of outages, and those outages can be between 24 and 48 time steps.

Those paramters are also available for the output mask. To use the exact same mask also as output there is an option `share_input_output_mask `. 

`train_dropout` is the option to remove sensors only from the training in- & output but not from validation or evaluation. 

So the overall idea is to see one set of masks as one experiment. This means that `--experiment_folder` should be a meaningful identifier for the experiment. 

### Usage of masks

For the execution, there is no longer the `--experiment_id` option available. This was replaced by `--mask_name` and `--mask_iter`. The first defines which `{experiment_folder}`-mask should be loaded and `--mask_iter` is used to use different masks for the N=3 runs that we have to do for each experiment/model combination. 


I hope this gives a brief overview how to use the masks. Please let me know if you find any bugs or bad behavior. Or also when you think we should add more dropout possibilities to model our experiments, which we can hopefully discuss next week.

**Comment by ywoelker** [comment]
I moved the application of the masks from the initialization of the dataloader class to the loading function. The idea is also to mask out the metadata accordingly. Also, the current x_mask as well as the label_mask are now class variables to the dataloader. This allows engines from baselines that require the mask (as OPCR for confidence intervals) to insert them into the model call. I decided for a class variable to not blow up the forward method call with two additional paramters.

Related commit: 61e81c77c27fd1e3c09092b054c
